### DISCUSSION TITLE
Sharing a quick Jupyter notebook file to demonstrate use right in a browser

### DISCUSSION THREAD

**Discussion by fomightez** [ialbert/bio#26] [Show and tell]
Jupyter notebook file to demonstrate use right in a browser can be [run by clicking here](https://mybinder.org/v2/gist/fomightez/8015b8455e2360e1674b8af6e3821d0a/HEAD?urlpath=%2Flab%2Ftree%2FDemonstrate_Bio_package.ipynb).

[Static version](https://nbviewer.org/urls/gist.githubusercontent.com/fomightez/8015b8455e2360e1674b8af6e3821d0a/raw/8d423530e091ed48640dfd3e87d6fd2435a5dfbd/Demonstrate_Bio_package.ipynb). 

I'm assuming the `define` in the README became `explain`?

**Comment by ialbert** [comment]
yes, probably it would be a good idea to run these example as a notebook as you do, that way errors like these get caught immediately.

could you create a command that dumps a binder into an HTML file?

**Reply by fomightez** [reply]
Here's commands you can run in that session that will freshly run the notebook in the launched session and save both markdown and HTML versions of the code. Is that what you meant?

```shell
pip install jupytext
jupytext --execute Demonstrate_Bio_package.ipynb; echo "[jupytext] Done executing notebook and saving freshly executed version."
jupytext --to md Demonstrate_Bio_package.ipynb
jupyter nbconvert --to html Demonstrate_Bio_package.ipynb
```
Combining the commands and output with Github actions or another script, the output from those commands could be used as input to automated tests and even building the recently run examples into the documentation or README.

**Reply by ialbert** [reply]
yes it is kind of what I was asking, 

would it be possible to generate that HTML output from a markdown file, rather than an ipython notebook?

**Reply by fomightez** [reply]
Yes, Jupytext can run markdown. As an example I'll take `Demonstrate_Bio_package.md` that my last code block would make and remove the Jupytext-added header to get a more basic markdown file and then run the code in the code blocks using Jupytext. Finally, I'll convert the output from running the markdown to HTML:

```shell
tail -n 38 Demonstrate_Bio_package.md > simple_md_with_code.md
jupytext --to notebook --execute simple_md_with_code.md
jupyter nbconvert --to html simple_md_with_code.ipynb
```

Perhaps an issue with what you are planning is that exclamation points are needed to run command line-based stuff in the typical-python based kernel. You can actually install a Bash kernel and tell Jupytext to use that, like so:

```shell
pip install bash_kernel==0.7.1
python -m bash_kernel.install
jupytext --to notebook --set-kernel bash --execute simple_md_with_code.md
```
That runs the `bio` command line commands without the need for the exclamation point.

**Reply by ialbert** [reply]
thanks for the explanations, I will add this to the todo list
